API: Optimize np.isin and np.in1d for integer arrays and add `kind=`

MilesCranmer opened the pull request:
This commit adds some simple functionality to np.isin and np.in1d that greatly increases performance when both arrays are integral. It works by creating a boolean array with elements set to 1 where the parent array (ar2) has elements and 0 otherwise. This array is then indexed by the child array (ar1) to create the output.

MilesCranmer committed:
MAINT: Optimize np.isin for integer arrays

- This optimization indexes with an intermediary boolean array to speed up
numpy.isin and numpy.in1d for integer arrays over a range of optimal parameters
which are calculated.

MilesCranmer commented:
Posted this to the mailing list as well

hameerabbasi commented:
I believe the flag shouldn't be necessary at all. Your optimisation should simply work if both dtypes are integral. [This shows](https://stackoverflow.com/a/22476290/774273) how to check for that.

adeak commented:
Have you done any checks to see what happens to performance when there are, say, a small number of large integers? I imagine at one point the cost of creating a huge and very sparse array might become noticeable. Of course I'd imagine such cases (assuming they exist) to be rare in nature.

MilesCranmer commented:
Thanks @hameerabbasi, I will make it automatic.

@adeak you make a good point. I'll make some benchmarks on my machine and post them. I suppose we can use the metric float(np.max(ar2) - np.min(ar2))/ar2.size to decide whether to apply this optimization?

MilesCranmer commented:
Here are some benchmarks: https://docs.google.com/spreadsheets/d/1vAx_zxQqqOTXOIFRMtafpUGnd-hquYOqQpxPa7JYR4c/edit?usp=sharing

(Red color shows when the normal version is twice as fast as the new version)

Suprisingly even in the very rare case of (np.max(ar) - np.min(ar)) = 10, but 10^10 elements, this version of np.isin() only takes 5 seconds on my machine. So maybe we can just take this for all integer arrays? And if one wants to use the regular algorithm, they can set it to be floating type. Thoughts?

adeak commented on numpy/lib/arraysetops.py:
```diff
@@ -519,6 +522,27 @@ def in1d(ar1, ar2, assume_unique=False, invert=False):
     ar2 = np.asarray(ar2).ravel()
 
+    # Make sure we won't encounter memory errors
+    if fast_integers and np.max(ar2) - np.min(ar2) < 8e8:
```
I counted four occurences of `np.min(ar2)` and three of `np.max(ar2)`. It would probably make sense not to look up the extrema each time.
Also I keep forgetting: is there any difference between `np.max(ar2)` and `ar2.max()`? Does the former merely call the latter?

MilesCranmer replied:
Good point. And yes if I recall correctly it is just an alias.

mattip replied:
> is there any difference between np.max(ar2) and ar2.max()? Does the former merely call the latter?

Short answer - they both call the `umath` `ufunc` `numpy.maximum.reduce` via python calls, but via different paths. Long answer:

- The first one calls [`_wrap_reduction(ar2, numpy.maximum, ...)`](https://github.com/numpy/numpy/blob/v1.15.2/numpy/core/fromnumeric.py#L2333) which checks if ar2 is an ndarray, then calls [`numpy.maximum.reduce(ar2, ...)`](https://github.com/numpy/numpy/blob/v1.15.2/numpy/core/fromnumeric.py#L83).

- The second maps to [`array_max`](https://github.com/numpy/numpy/blob/v1.15.2/numpy/core/src/multiarray/methods.c#L317) in C, which ends up calling out to python in [`_methods.py`](https://github.com/numpy/numpy/blob/v1.15.2/numpy/core/_methods.py#L17) and calling `numpy.maximum.reduce(ar2, ...)`.

MilesCranmer commented:
Sorry, I edited my above comment - instead I meant the worst performance is for 10 elements with a range of 10^10. Perhaps we should add some metric to bypass this algorithm in cases like this after all?

MilesCranmer commented:
So in looking at the benchmarks where the original algorithm is not twice as good as the new one, I see the following figure. The y-axis is log10(array size) and the x-axis is log10(range) for the benchmarks. The color is the log10(speedup) - i.e., a value of 1 for color means the new code is 10x faster.

<img width="629" alt="screen shot 2018-10-01 at 5 42 15 pm" src="https://user-images.githubusercontent.com/7593028/46317552-63d35180-c5a1-11e8-8ec8-0869c274245f.png">


The gap looks to have a linear top, aside from the rightmost column - I'm not sure what is going on there - maybe a memory problem? This is all pretty approximate and specific to my own machine so I think it is okay to be approximate here - we really just want to prevent the extreme slow behaviour from this addition.

To block out the space where the old code is <2x faster than the new code, the line:

log10(size) = (log10(range) - 2.27) / 0.927

cuts it appropriately. Therefore, I can add this check for whether the new code or old code should run.

MilesCranmer commented:
With the conditional run on array parameters above, here is a histogram of the log10(speedups) for all benchmarks. As you can, the vast majority of settings result in speedups and there are no harsh slowdowns anywhere.

<img width="629" alt="screen shot 2018-10-01 at 6 00 54 pm" src="https://user-images.githubusercontent.com/7593028/46318324-1b696300-c5a4-11e8-80ed-4dc71377b3e7.png">

MilesCranmer commented:
Added all suggestions. Please let me know what you think.

Cheers,
Miles

hameerabbasi reviewed (commented):
Some other feedback. I still have some concerns about the memory consumption of this algorithm versus the original one, but I won't stop this moving forward based on those.

MilesCranmer commented:
Thanks for the suggestions, I will implement them.

Regarding memory: the memory usage is already capped at 8x10^8 elements for the helper array before using the normal method so 800 MB max. To get to that point, you would need greater than:

size >= 10**((log10(range) - 2.27) / 0.927) = 14,303,321

So, minimum 14,303,321 elements in ar2 for it to be within the optimal parameter cutoff. In order to have values all the way up to 8x10^8 elements, you would need at least 32-bit integers, so your input array is minimum 58 MB already. So I think you would already be dealing with memory-heavy calculations in that case so maybe it is expected. What do you think?

hameerabbasi commented:
> 800 MB

Do you have a comparison versus the original algorithm possibly? That would be more helpful. I'm jut commenting with general tips, I know nothing of the algorithms used.

In case your algorithm is worse: 800 MB might not be much on today's multi-gigabyte-memory desktops, but we have to keep in mind IoT devices and things like Raspberry Pis as well, which is the use-case I'm more concerned about.

MilesCranmer commented on numpy/lib/arraysetops.py:
```diff
@@ -519,6 +519,45 @@ def in1d(ar1, ar2, assume_unique=False, invert=False):
     ar2 = np.asarray(ar2).ravel()
 
+    # Check if we can use a fast integer algorithm:
+    integer_arrays = (np.issubdtype(ar1.dtype, np.integer) &
```
Thanks, I'll switch to `and`.

I am not sure the boolean types will work - they would require a conversion to uint8 (should I do this?) since using booleans as an index would be interpreted as a mask.

hameerabbasi replied:
No, you would just write `condition1 and condition2` instead of `condition1 & condition2`. Another benefit of this is short-circuiting, which I forgot earlier.

hameerabbasi replied:
> I am not sure the boolean types will work - they would require a conversion to uint8 (should I do this?) since using booleans as an index would be interpreted as a mask.

You could probably use `arr.view(np.uint8)` for a zero-copy view, along with adding a check for `dtype == np.bool_` for each of the arrays. I guess you could just do `if arr.dtype == np.bool_: arr = arr.view(np.uint8)`.

hameerabbasi replied:
Not really. :-) It just creates a "view" into the existing data. https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.view.html

hameerabbasi replied:
> (I suppose it wouldn't really make sense to run np.isin on a boolean/integer ar1/ar2, respectively?)

Try mirroring the existing behaviour, check what that is by running it once.

MilesCranmer replied:
Oh sorry what I mean is: I think both ar1.dtype and ar2.dtype should be bool before applying this np.uint8 conversion. 

If not, one could get an error when running np.isin([3, 1], [False, True, False]). The correct return I think should be [False, False]. However, if we only check one array at a time for boolean and convert it to uint8,  the second array would be converted to unit8 while the first is kept the same, and then [False, True] would be returned.

So requiring both dtypes to be np.bool_ fixes this issue.

MilesCranmer replied:
Adding this now.

MilesCranmer commented:
So from another benchmark, it looks like this new algorithm also does better for memory usage than the traditional method, as least for one case.

In the same google sheet (Shee name: "Worst case memory usage") : [(here)](https://docs.google.com/spreadsheets/d/1vAx_zxQqqOTXOIFRMtafpUGnd-hquYOqQpxPa7JYR4c/edit?usp=sharing), I show the memory usage as recorded by the package memory_profiler: https://pypi.org/project/memory_profiler/. The memory is given every 0.1 seconds (I add sleep(.1) at major steps in the code to increase the resolution). 

Basically, for parameters:

range ~ 8x10^8 - 1
size ~ 5x10^6,

I record, for the normal algorithm:

max memory usage = 1.8 GB

and the fast algorithm:

max memory usage = 1.4 GB.


So perhaps I should just turn the range < 8e8 requirement off?

hameerabbasi commented:
> So perhaps I should just turn the range < 8e8 requirement off?

Sounds good to me, thanks for being so thorough. I appreciate your patience with my review. After my remaining comments are addressed, this gets a +1 from me.

MilesCranmer committed:
MAINT: Optimize np.isin for boolean arrays

- This change converts boolean input to numpy.isin and numpy.in1d into uint8, allowing
the function to use the new optimized code

MilesCranmer commented:
All requests have been added. Thanks for your help @hameerabbasi!

Cheers,
Miles

hameerabbasi reviewed (approved):
Great, I'd make sure the tests hit the new area of the code, both with `bool` data types and one integer data type. Other than that, this is a +1 from me.

MilesCranmer commented:
The tests actually already hit the new functionality with their integer tests. I will add some integer tests which explicitly hit the old code (which requires range > 186 x 10^(0.927 x size) ) as well as boolean tests which don't exist yet.

MilesCranmer committed:
TST: add tests for in1d/isin

- These tests hit the areas of the old algorithm which are now replaced
for the majority of cases
- Another test tests behaviour of boolean input to isin/in1d

MilesCranmer commented:
Added suggestions and tests.

hameerabbasi reviewed (approved):
This gets a +1 from me -- Thanks @MilesCranmer for the PR.

MilesCranmer commented:
Anything else I need to do/people I need to email to have this merged?

Thanks,
Miles

hameerabbasi commented:
I don't have write access. Perhaps ping the mailing list?

mattip commented:
Does this have any impact on existing asv benchmarks?

hameerabbasi commented on numpy/lib/arraysetops.py:
```diff
@@ -519,6 +519,50 @@ def in1d(ar1, ar2, assume_unique=False, invert=False):
+                                                        ar2_min]
+
+            if invert:
+                return np.invert(outgoing_array)
```
I would perhaps replace this line with `np.invert(outgoing_array, out=outgoing_array)` without the `return`.

hameerabbasi reviewed (approved):
@mattip Makes a valid point about the benchmarks, if there aren't any asv benchmarks with `np.isin`, it would be nice to add them. This has my approval either way though: Don't want to make the barrier to entry too high.

MilesCranmer commented:
I just looked and I couldn't find any usage of isin or in1d in the benchmarks.

hameerabbasi commented:
Perhaps add some, benchmarks in newer versions can also be run against old commits with `asv`.

shoyer commented:
Currently, `np.isin()` either uses sorting or a naive loop for the implementation.

Am I right in interpreting this PR as adding an equivalent of radix sort?

hameerabbasi commented:
> Am I right in interpreting this PR as adding an equivalent of radix sort?

IIUC, it's a counting sort.

mattip commented:
It seems strange to me to merge a performance-related PR with no before-and-after benchmarks.

hameerabbasi commented:
> It seems strange to me to merge a performance-related PR with no before-and-after benchmarks.

There are a few almost at the top. Not absolute, but relative numbers.

mattip commented:
How do I run those?

hameerabbasi commented:
> How do I run those?

They aren't in the repo yet. I believe @MilesCranmer will add ASV benchmarks later this week.

charris commented:
Could you add a release note under `Improvements`?

mattip commented:
@MilesCranmer we are still waiting for the benchmark additions and a mention in the release notes. Would you like help with that?

MilesCranmer commented:
Hi @mattip, sorry for the delay - I forgot about this PR. I'll add this to my todo list for this weekend!
Cheers,
Miles

MilesCranmer committed:
BENCH: Add benchmark for integer input to np.isin

MilesCranmer committed:
DOC: Add release notes for faster np.isin

MilesCranmer commented:
I added benchmarks and some release notes. Please let me know what you think.

Here is the speedup from a run on my computer:

<img width="1417" alt="screen shot 2018-12-25 at 7 08 30 pm" src="https://user-images.githubusercontent.com/7593028/50427885-924e8900-087f-11e9-839d-74bee177c626.png">

MilesCranmer commented:
I'm not sure why GitHub shows those other commits here (7551286-
0675456). I am not tagged in the original commits in the actual git log - GitHub displays me on them for some reason.

Is that a known GitHub bug?

mattip commented:
> GitHub shows those other commits here

You are working off a branch named master and want to merge to that same branch. That is OK, until you `rebase` before you `pull` from `origin`. You might be able to use `git rebase -i HEAD~12` and then `drop` the commits that are not part of your PR, you will then need to force push again.

MilesCranmer commented:
Okay, thanks, my mistake. I'll manually patch the commits on top of a clean repo and also change the "ENH" to "MAINT" in the commit text.

MilesCranmer commented:
I fixed the commit history, sorry for my mistake earlier. Let me know if you think this is ready to merge.

Cheers,
Miles

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -551,6 +551,50 @@ def in1d(ar1, ar2, assume_unique=False, invert=False):
+    if integer_arrays:
+        ar2_min = np.min(ar2)
+        ar2_max = np.max(ar2)
+        ar2_range = ar2_max - ar2_min
```
This can overflow if `ar2_min == np.int8(-100)`, `ar2_max == np.int8(100)`. Does the algorithm work anyway?

MilesCranmer replied:
It seems it actually works in this case but it does spit out warnings:
```
>>> import numpy as np
>>> y = np.array([-100, 5, 100]).astype(np.int8)
>>> x = np.array([5, 50, 100]).astype(np.int8)
>>> np.isin(x, y)
array([ True, False,  True])
```

With warning:
```
/opt/conda/lib/python3.7/site-packages/numpy-1.17.0.dev0+5d12e37-py3.7-linux-x86_64.egg/numpy/lib/arraysetops.py:568: RuntimeWarning: overflow encountered in byte_scalars
  ar2_range = ar2_max - ar2_min
/opt/conda/lib/python3.7/site-packages/numpy-1.17.0.dev0+5d12e37-py3.7-linux-x86_64.egg/numpy/lib/arraysetops.py:576: RuntimeWarning: invalid value encountered in log10
  ((np.log10(ar2_range + 1) - 2.27) / 0.927)
```

It looks like the optimal_parameters flag catches this invalid value and the old algorithm is used instead. Should I check for this or leave it as is?

MilesCranmer replied:
I check for overflow explicitly in 40ad85e. Let me know what you think.

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -551,6 +551,50 @@ def in1d(ar1, ar2, assume_unique=False, invert=False):
+                                                        ar2_min]
+
+            if invert:
+                np.invert(outgoing_array, out=outgoing_array)
```
You could probably do better here by swapping the `np.zeros` above for `np.ones` if invert is specified

MilesCranmer replied:
Agreed, I'll add this.

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -551,6 +551,50 @@ def in1d(ar1, ar2, assume_unique=False, invert=False):
+    integer_arrays = (np.issubdtype(ar1.dtype, np.integer) and
+                      np.issubdtype(ar2.dtype, np.integer))
+
+    # Make sure we won't encounter memory errors
```
I'm not sure I understand the intent behind this comment.

MilesCranmer replied:
Actually this comment should be removed as I discovered the new algorithm actually lowers memory use (contrary to my assumption).

eric-wieser commented on numpy/lib/tests/test_arraysetops.py:
```diff
@@ -289,6 +289,75 @@ def test_in1d(self):
 
+    def test_in1d_hit_alternate_algorithm_floats(self):
+        # Perform the above test but with floats
+        # This forces the old in1d (pre-Oct 1/2018) algorithm
```
It would be nice to have some clearer way to force one algorithm or the other, just so that we can use the same contrived corner-cases for each in tests.

MilesCranmer replied:
Should I add some debug argument to isin/in1d for this? I agree with your comment.

MilesCranmer replied:
Added a flag to isin/in1d for this and extended the tests.

hameerabbasi replied:
It would be nice if you could make it a 3-way switch. `None` auto detects, `True`/`False` force one or the other.

eric-wieser commented on numpy/lib/tests/test_arraysetops.py:
```diff
@@ -320,6 +396,33 @@ def test_in1d_ravel(self):
+        """Hit the standard isin code with integers"""
+        # Need extreme range to hit standard code
+        a = np.array([5, 4, 5, 3, 4, 4, 1e9])
+        a = a.astype(np.int64)
```
Can combine these two lines as `np.array(..., dtype=np.int64)`

MilesCranmer commented:
Resolved some comments; waiting for additional feedback on others.

MilesCranmer committed:
MAINT: Check for overflow in integral np.isin

- Before this fix, if, e.g., int8(-100) and int8(100) are the min and
max, respectively, of ar2, ar2_range would experience an integer
overflow and optimal_parameters would be an invalid value.
``if optimal_parameters:`` would skip the fast integer algorithm
regardless, but warnings would be generated when they are unnecessary.
- np.log10(... + 1) was changed to np.log10(... + 1.0) to avoid further
potential problems from overflow here.
- See https://github.com/numpy/numpy/pull/12065 for discussion.

MilesCranmer committed:
TST: Extend np.in1d tests to old algorithm

- Add flag ``_slow_integer`` to np.isin/np.in1d to force the use of the
old isin/in1d algorithm for integers.

hameerabbasi commented on numpy/lib/arraysetops.py:
```diff
@@ -564,17 +564,27 @@ def in1d(ar1, ar2, assume_unique=False, invert=False):
-            )
-
+        # Check for integer overflow
+        old_error_settings = np.seterr(over='raise')
```
You can use

```python
with np.seterr(...):
    code here
```

which will auto-revert in all failing cases.

MilesCranmer replied:
Done (with np.errstate)

MilesCranmer commented:
Thanks @hameerabbasi, adding these now. Do you know why the build is failing? It builds and tests (the isin parts) fine on my machine.

I see another implementation of isin/in1d in ma/extras.py - is this the other signature that the travis build complains about?

eric-wieser commented on numpy/lib/arraysetops.py:
```diff
@@ -551,6 +555,60 @@ def in1d(ar1, ar2, assume_unique=False, invert=False):
+                # https://github.com/numpy/numpy/pull/12065
+                optimal_parameters = (
+                        np.log10(ar2_size + 1.0) >
+                        ((np.log10(ar2_range + 1.0) - 2.27) / 0.927)
```
Why the +1?

MilesCranmer replied:
For `ar2_range` of 0 it is a way to prevent blow-up. Then I add a 1 to the other log10 to balance out the original fit. But it doesn't change the "optimal region" that much.

eric-wieser replied:
+1 makes complete sense for the 'ar2_range', since you use '+1' on it below when using it too. Would be nice to eliminate the '+1' on the size, since it doesn't have any reasonable interpretation.

MilesCranmer replied:
Okay sounds good.

hameerabbasi commented:
@MilesCranmer Above the `isin` function there is a `_isin_dispatcher`, you need to change that to match the `isin` function. Remember that it must have the exact same signature, but all defaults (where applicable) should be `None`. It should return an iterable of all `array_like` arguments to the function.

In your case, you just need to add the extra kwarg to the dispatcher.

MilesCranmer commented:
Would this be ready to merge?

MilesCranmer commented:
Hi, how is this going? Should I resolve the new conflicts?

shoyer commented:
It would be interesting to see how the performance numbers change here after we add radix sort (https://github.com/numpy/numpy/pull/12586). Is this still worth doing?

MilesCranmer commented:
I encourage someone to double check, but I just pulled #12586 and this addition does still offer significant improvements for large arrays. 

Without this addition (improvement over v1.15.0):

<img width="643" alt="screen shot 2019-01-31 at 8 27 39 am" src="https://user-images.githubusercontent.com/7593028/52057124-2f14de00-2532-11e9-8e78-55793a5a2655.png">

(the (x, y) tuple is x=size, y=range)

With this addition:

<img width="637" alt="screen shot 2019-01-31 at 8 22 58 am" src="https://user-images.githubusercontent.com/7593028/52057108-27553980-2532-11e9-84ed-492804318043.png">

hameerabbasi commented:
The issue is that with the backwards compatibility problem pointed out, radix sort may not (read: is unlikely to) become the default at all. The resolution to that is yet to be determined.

I'm +1 for this going in either way, after a rebase. I'd have pushed the button myself a while ago if I had the permissions. :-)

There's a sweet spot between Facebook's "move fast and break things" and NumPy's current "backwards compatibility over all else" that I think we need to reach.

seberg commented:
We discussed this in todays meeting and think we should try to push this forward probably.  One thing we noted is that the "private" keyword argument seems odd (and unprecedented in NumPy).  So the option would be to either remove it (if we need absolutely need it, create a private helper?), or find a better name (`optimize=`, `method=`, I am not sure?).

At this point, I am also unsure about the exact trade-offs we are making here, e.g. when the min-max difference leads to the old algorithm.

In general, I guess this PR probably was stalled on more detailed review, unfortunately.

charris commented:
@MilesCranmer Needs a rebase. The note also needs to be in the new `doc/release/upcoming_changes` directory.

MilesCranmer committed:
DOC: Move release notes for faster np.isin

MilesCranmer committed:
MAINT: Fix linting errors in in1d tests

MilesCranmer commented:
Rebase done, and fixed some linting issues. Let me know what I need to do next.

MilesCranmer commented:
@seberg I like your suggestion to add a `method` argument to force `isin` to use the old (current) algorithm. What do others think?

MilesCranmer committed:
DOC: Undo change to old release notes

MilesCranmer committed:
MAINT: Change `_slow_integer` parameter to `method`

MilesCranmer committed:
Add check for methods

MilesCranmer committed:
MAINT: Update tests to use new `method` argument

MilesCranmer committed:
DOC: Specify constraints of method in docstring

MilesCranmer committed:
MAINT: Fix linting error in test

MilesCranmer committed:
DOC: Describe default in docstring

MilesCranmer commented:
Ready for review!

I took @seberg's suggestion of using `method`, which is described below (the docstring for `in1d` and `isin`).
```
    method : {'auto', 'sort', 'dictionary'}, optional
        The algorithm to use. This will not affect the final result,
        but will affect the speed. Default is 'auto'.

        - If 'sort', will use a sort-based approach.
        - If 'dictionary', will use a key-dictionary approach similar
          to a radix sort. This is only available for boolean and
          integer arrays.
        - If 'auto', will automatically choose the method which is
          expected to perform the fastest, which depends
          on the size and range of `ar2`. For larger sizes,
          'dictionary' is chosen. For larger range or smaller
          sizes, 'sort' is chosen.
```

adeak commented on doc/release/upcoming_changes/12065.performance.rst:
```diff
@@ -0,0 +1,8 @@
+----------------------------------------------------------------
+``np.in1d`` (used by ``np.isin``) can now switch to a faster algorithm
+(up to >10x faster) when it is fed two integer arrays.
+The algorithm bares similarities to a counting sort in that it
```
```suggestion
The algorithm beares similarities to a counting sort in that it
```

adeak commented on doc/release/upcoming_changes/12065.performance.rst:
```diff
@@ -0,0 +1,8 @@
+----------------------------------------------------------------
+``np.in1d`` (used by ``np.isin``) can now switch to a faster algorithm
+(up to >10x faster) when it is fed two integer arrays.
+The algorithm bares similarities to a counting sort in that it
```
This compares to a counting sort, but the docstring compares to a radix sort. I don't have any formal CS background so either or both can be correct, just pointing this out in case you want to make the same comparison in both places.

hameerabbasi replied:
I think from the description, counting sort is closer, but I'd be careful with that approach as if you feed it the max value of `uint64`, it could OOM, which leads me to believe that the actual implementation is similar to a radix sort.

adeak replied:
The current "new" implementation seems to explicitly allocate an auxiliary array of size `arr2.ptp()` (give or take), so the potential OOM issue is probably present.

hameerabbasi replied:
> actual implementation is similar to a radix sort.

I just checked the implementation: It's a counting sort, with an automatic switch-over.

MilesCranmer replied:
Fixed the docstring, thanks.

adeak commented on doc/release/upcoming_changes/12065.performance.rst:
```diff
@@ -0,0 +1,8 @@
+``np.in1d`` (used by ``np.isin``) can now switch to a faster algorithm
+(up to >10x faster) when it is fed two integer arrays.
+The algorithm bares similarities to a counting sort in that it
+uses the ``test_elements`` argument to index a boolean helper
```
The `test_elements` and `elements` kwargs are defined on `isin()`, but this paragraph technically talks about `in1d()`. I don't know if this can be confusing for anybody (i.e. that there is no `test_elements` parameter for `in1d()`, because it calls its parameters `ar1` and `ar2`).

MilesCranmer replied:
Good point. I cleaned this up and also mentioned some things about the memory allocation being different.

adeak commented on numpy/lib/arraysetops.py:
```diff
@@ -593,6 +607,70 @@ def in1d(ar1, ar2, assume_unique=False, invert=False):
+            return outgoing_array
+    elif method == 'dictionary':
+        raise ValueError(
+            "'dictionary' method is only supported for non-integer arrays. "
```
I think this was meant to say
```suggestion
            "'dictionary' method is only supported for boolean or integer arrays. "
```

MilesCranmer replied:
Thanks, fixed.

adeak commented on numpy/lib/arraysetops.py:
```diff
@@ -544,6 +545,19 @@ def in1d(ar1, ar2, assume_unique=False, invert=False):
         False where an element of `ar1` is in `ar2` and True otherwise).
         Default is False. ``np.in1d(a, b, invert=True)`` is equivalent
         to (but is faster than) ``np.invert(in1d(a, b))``.
+    method : {'auto', 'sort', 'dictionary'}, optional
+        The algorithm to use. This will not affect the final result,
+        but will affect the speed. Default is 'auto'.
```
These remarks make it sound a bit as if the only difference were speed. If I understand correctly, with the new method we're allocating an auxiliary array as large as the range of the second array, although with a 1-byte dtype. I'm not familiar with the memory need of the old method, but can it happen that the new method uses considerably more memory than the old one, which can affect users?

If I understand correctly the default `'auto'` should choose the old method for large ranges, so this probably only matters if a user goes out of their way to ask for the new method with `method='dictionary'`.

MilesCranmer replied:
The criterion for choosing the new method is:
$$ \log_{10}(\text{size}) > \frac{\log_{10}(\text{range}) - 2.27}{0.927}$$
which is roughly equivalent to:
$$ \text{size} > \text{range}^{1.079} / 281 $$
where `size` and `range` are of ar2. The size of the newly created matrix will have size equal to `range`. _(That_ limit is purely based on performance rather than memory.)

Therefore, in the worst case, the newly created counting array could have up to 300x the number of elements of the input array, with a 1-byte dtype.

Do you think this limit should be smaller by default? Any idea what the memory usage is for the old method?

hameerabbasi replied:
> Therefore, in the worst case, the newly created counting array could have up to 300x the number of elements of the input array, with a 1-byte dtype.

I think for embedded applications it could make a huge difference.

adeak replied:
To be clear I'm mostly asking about the docs: I think we should make it clear that the new method uses an auxiliary uint8 array as large as `arr2.ptp()`.

The criterion you posted here says that for "huge" range we need "even huger" size for `ar2`. I'd be more concerned about a "small" `ar2` array with "huge" values inside. After all, as @hameerabbasi noted, the worst-case size is approximately `np.iinfo(np.uint64).max == 18446744073709551615`. That's a lot, and independent from the size of the input arrays.

Considering this it might be wise to set a hard range cap on the `'auto'` method as well, so as not to trip up old code. But I'd leave this to the NumPy devs, I'm just the peanut gallery :)

MilesCranmer replied:
Let me know what you think would be a good constraint based on memory usage and I'm very happy to implement it. The inequality I posted above is purely based on performance considerations, not memory.

Also happy to describe this in the docs as well, just let me know what is decided on.

MilesCranmer replied:
> The criterion you posted here says that for "huge" range we need "even huger" size for ar2. I'd be more concerned about a "small" ar2 array with "huge" values inside. 

Not quite: it says that for ar2 having range $r$, you need ar2's size to be at least $0.0035$ times as big for the algorithm to be used (roughly). Then, the intermediate array will have size equal to $r$.

MilesCranmer replied:
A safer option might be to only create an intermediate array if it has the same size as the input arrays (rather than currently: 300x the input array size). Then at worst you are doubling memory which seems fine. Thoughts?

In the docstring I can specify `"dictionary"` method is usually much faster if the range is not too much bigger than input array sizes, but the memory usage will be equal to your data range.

MilesCranmer replied:
The old method, by eye, seems to have 5x to 10x the memory usage of the input arrays via intermediate allocations. Those limits could be used for complete consistency with the old method if you would like?

adeak replied:
> The old method, by eye, seems to have 5x to 10x the memory usage of the input arrays via intermediate allocations. 

But not using all that memory at once, right? I would assume that a single 1 GB allocation is a lot worse than ten 100 MB allocations, because the latter can reuse memory that's just been freed.

MilesCranmer replied:
Those variables look to be in-scope so I don't think they are freed until the function is complete, unless you re-assigned their variable names. Right? (I am not too familiar with how Python's garbage collection works)

adeak replied:
Yeah, if all those intermediate arrays have references (by being bound to individual names) then they stick around until the end of the function. I thought most of these were intermediate arrays in a chain of calculations that aren't even bound to a nem  (I haven't looked at the old implementation yet, sorry.)

MilesCranmer replied:
I will resolve this and summarize the discussion in the main thread, and ask for input.

hameerabbasi commented on numpy/lib/arraysetops.py:
```diff
@@ -593,6 +607,70 @@ def in1d(ar1, ar2, assume_unique=False, invert=False):
+    integer_arrays = (np.issubdtype(ar1.dtype, np.integer) and
+                      np.issubdtype(ar2.dtype, np.integer))
+
+    if method not in ['auto', 'sort', 'dictionary']:
```
Optional nit:
```suggestion
    if method not in {'auto', 'sort', 'dictionary'}:
```

hameerabbasi commented on numpy/lib/arraysetops.py:
```diff
@@ -593,6 +607,70 @@ def in1d(ar1, ar2, assume_unique=False, invert=False):
+            "Invalid method: {0}. ".format(method)
+            + "Please use 'auto', 'sort' or 'dictionary'.")
+
+    if integer_arrays and method in ['auto', 'dictionary']:
```
Again:
```suggestion
    if integer_arrays and method in {'auto', 'dictionary'}:
```

MilesCranmer committed:
MAINT: Fix use of dispatcher for isin

MilesCranmer committed:
MAINT: Fix error message

MilesCranmer committed:
DOC: Fix text in changelog

Co-authored-by: Andras Deak <adeak@users.noreply.github.com>

MilesCranmer committed:
MAINT: Formatting changes for in1d

Co-authored-by: Hameer Abbasi <2190658+hameerabbasi@users.noreply.github.com>

MilesCranmer committed:
DOC: Improve docstring explanation

MilesCranmer committed:
MAINT: bool instead of np.bool_ dtype

MilesCranmer commented:
Maintainers, what would you recommend as a bound for the max size of an intermediate array **relative to the input array size**?

Performance-wise, this method is more efficient than the old method even when the memory allocation is ~300x larger than the input array. However, that's obviously a lot of memory to allocate by default, so we could have a more conservative default.

The old method uses 11x the memory of the input arrays for intermediate variables. Is 11x a good cutoff to implement, or could larger be used by default?

MilesCranmer commented:
Here is the calculation to arrive at 12x the memory for the old method:
```python
    # Say ar1, ar2 both size N.
    ar = np.concatenate((ar1, ar2))
    # ^2N
    order = ar.argsort(kind='mergesort')
    # ^2N
    sar = ar[order]
    # ^2N
    if invert:
        bool_ar = (sar[1:] != sar[:-1])
        # ^2N
    else:
        bool_ar = (sar[1:] == sar[:-1])
        # ^2N
    flag = np.concatenate((bool_ar, [invert]))
    # ^2N
    ret = np.empty(ar.shape, dtype=bool)
    # 2N
    ret[order] = flag
    # = 2N (ar) + 2N (order) + 2N (sar) + 2N (bool_ar) + 2N (flag) + 2N (ret) = 12N
```
Since these variables aren't explicitly re-assigned, they stay in memory until the function returns.

The new method's memory allocation is determined by the **range** of `ar2`, which is why `method="auto"` will switch to the old method if this allocation is too large. But it looks like the new method might also have a smaller memory allocation for small ranges of data.

MilesCranmer committed:
DOC: Clean up isin docstring

MilesCranmer committed:
DOC: Describe memory considerations in in1d/isin

MilesCranmer committed:
DOC: Clean up change log

MilesCranmer committed:
MAINT: Add back in1d tests of old method

MilesCranmer committed:
MAINT: Fix misplaced default in in1d test

MilesCranmer commented:
By the way, the speed-up is still looking just as good as it did ~4 years ago which is good to see:
<img width="932" alt="Screen Shot 2022-06-10 at 8 46 54 PM" src="https://user-images.githubusercontent.com/7593028/173166024-e173fb9b-d041-47f9-a458-5e17df22668d.png">

seberg commented:
Sorry, I have not kept up with the development here.  I know I initially brought up the "method", but if at all, I would prefer if we can avoid it.
12x the input array size is already pretty huge, I would not worry about bloating that up at least a bit more, but if we can just use that as a cut-off and it seems fine then that is also good?

Would you/the other reviewers who engaged be able to formulate a "recommendation" and see if that is just good enough to do without the `method`, but still a huge improvement?

Maybe we just need to spend a bit more time on going through the details, but any summary/recommendation would probably help.

MilesCranmer commented:
Hi @seberg,

Sure, I can use 12x as a cutoff for internal array allocations; otherwise, it will switch to the old algorithm.  I am unfamiliar with the memory budget that numpy would typically assume so I will rely on the wisdom of others. Is this the final change or is there anything else I need to do before the PR is merged?

Regarding the `"method"` flag, I am happy to change this. Please let me know your recommendation. (Out of curiosity, why remove it, as it is already an optional flag? I am not well-versed in internal numpy style guidelines, so maybe this is simply not standard to have a method-switching flag. I would certainly use the `"method"` flag, if available, as I don't have tight memory constraints, and `isin` is a performance bottleneck in some of my code.)

Cheers,
Miles

seberg commented:
I don't mind adding the kwarg much and a good/necessary thing will be to ping the mailing list.
(Reply probably is low there though, so the opinion of those involved here is most likely trump.)

However, any additional API is API we might regret adding at some point, if just because we don't like the `method` name, but prefer `kind` or because a new C implementation has no use for it.  Not adding the kwarg simply removes the API design dimension entirely and makes the decision easier.

Normally, I would say that a copy (or even two) is very normal for NumPy, with the exception of some in-place operations.  Since we are already at 12x, I don't mind a bit higher.  But if putting it at 12x or 20x is likely much the same... 12x seems fair.

I would be curious if @hameerabbasi or @adeak have a clear opinion/recommendation.

adeak commented:
I've been staying silent because I also don't know NumPy best practices, and I don't personally mind anything as long as we're in the same ballpark as originally. My only concern was the decoupling of the memory complexity from the _size_ of the input, but if we're setting a hard cap on the allocation size then this will naturally prevent accidental explosions of memory need with the default behaviour. So I'm good with whatever cap we choose as long as there is one.

MilesCranmer commented:
By the way, if `method` is removed as a kwarg, and the branch selection is forced to be automatic, is there a standard numpy debugging style for triggering specific branches in the unit-tests (e.g., global variable?)? Right now I explicitly set `method` in the unit-tests which allows me to confirm each method gets the correct `isin` output. 

I suppose this doesn't have to be done, but this means, for example, that the old method would never be unit-tested tested on boolean arrays, since those always default to the new method (as the range of data is small). I guess the old method would technically never be *used* on boolean arrays anyways, so maybe that's what we want for the unit-testing...?

shoyer commented:
I would strongly supporting adding a `method` or `kind` keyword argument, to allow users to custom this behavior. The speed vs memory tradeoffs discussed already make it clear that no single choice will be optimal for everyone.

jbrockmendel commented:
I've wondered for a while why numpy doesn't upstream the pandas implementation:

```
In [4]: x = np.random.randint(0, 100_000, size=10_000_000)

In [5]: y = np.random.randint(0, 100_000, size=10_000_000)

In [6]: %timeit np.in1d(x, y)
2.66 s ± 119 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [10]: %timeit pd.core.algorithms.isin(x, y)
246 ms ± 17.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

seberg commented:
Hmmm, it also means we have to create a full hashtable implementation?  To be fair... maybe it *is* time to create a basic hashtable implementation for NumPy...
I suppose a hashtable might fill the gap so well, that this approach is generally unnecessary...

For pandas making these types of things fast is at the very core, for us it is not really :).  But I guess that doesn't mean we shouldn't consider it.

MilesCranmer commented:
For what it's worth, the pandas one seems a bit slower than this new implementation:
<img width="698" alt="Screen Shot 2022-06-16 at 12 17 29 PM" src="https://user-images.githubusercontent.com/7593028/174117907-c62acf49-81af-41d9-8f25-61cb4c1a6909.png">

Maybe because it is creating a hashtable, whereas this is just using an array as the hashtable.

But perhaps generally you would want to fall back to the pandas one if you have non-integer data.

jbrockmendel commented:
> For what it's worth, the pandas one seems a bit slower than this new implementation:

Yah, the pandas implementation has a heuristic to fall back to the numpy version for some size combinations.  I don't know how that heuristic was chosen.  Wouldn't surprise me if there are regimes where all three versions perform well.

The hashtable idea isn't worth derailing this thread or holding up this PR.

MilesCranmer committed:
MAINT: Switch to old in1d for large memory usage

MilesCranmer commented:
I have implemented the suggested changes. Now, `method="auto"` performs the following: if the intermediate array created by the new method requires memory allocation that is more than 6 times the combined memory allocation of the input arrays, the old method will be used instead. But, `method="dictionary"` will force the new method regardless of required memory allocations.

shoyer commented:
As I wrote on the mailing list, I would lean towards this new parameter `kind` and using a default value of `kind=None` for automatic selection, for consistency with `np.sort`. I can even imagine supporting something like `kind='quicksort'` in the future, to explicitly select using the quicksort algorithm under the hood.

The only place where I found NumPy using the name `method` is for `np.quantile`, where it can actually change what values are calculate.

MilesCranmer commented:
Sounds good, will switch to `kind` and then `"mergesort"` for the name of the sort-based method instead of simply `"sort"`. (The error message would then inform that other sorts are not currently available).

MilesCranmer committed:
MAINT: Switch parameter name to 'kind' over 'method'

MilesCranmer committed:
TST: Use new "kind" argument over "method"

MilesCranmer committed:
MAINT: kind now uses "mergesort" instead of "sort"

MilesCranmer commented:
Done. Here's the updated docstring:

```
   kind : {None, 'mergesort', 'dictionary'}, optional
        The algorithm to use. This will not affect the final result,
        but will affect the speed. Default will select automatically
        based on memory considerations.

        - If 'mergesort', will use a mergesort-based approach. This will have
          a memory usage of roughly 6 times the sum of the sizes of
          `ar1` and `ar2`, not accounting for size of dtypes.
        - If 'dictionary', will use a key-dictionary approach similar
          to a counting sort. This is only available for boolean and
          integer arrays. This will have a memory usage of the
          size of `ar1` plus the max-min value of `ar2`. This tends
          to be the faster method if the following formula is true:
          `log10(len(ar2)) > (log10(max(ar2)-min(ar2)) - 2.27) / 0.927`,
          but may use greater memory.
        - If `None`, will automatically choose 'dictionary' if
          the required memory allocation is less than or equal to
          6 times the sum of the sizes of `ar1` and `ar2`,
          otherwise will use 'mergesort'. This is done to not use
          a large amount of memory by default, even though
           'dictionary' may be faster in most cases.
```

seberg commented on numpy/lib/arraysetops.py:
```diff
@@ -593,6 +615,75 @@ def in1d(ar1, ar2, assume_unique=False, invert=False):
+                    ar2_range <= 6 * (ar1_size + ar2_size)
+                )
+            except FloatingPointError:
+                below_memory_constraint = False
```
Just a thought:  Maybe move this up and add a `finally:` clause?  That would make it a bit more clear where the overflow would occur.

seberg replied:
I suppose this needs an extra error path?  If the method is forced, but is not compatible with the input range?

MilesCranmer replied:
Which code do you propose to move to the `finally:` clause?

MilesCranmer replied:
I added a `range_safe_from_overflow` parameter that will forcefully skip the `"dictionary"` method if there's an overflow. Should I add an error if it gets skipped despite explicitly setting it, or just silently do so?

seberg replied:
The `ar2_range <= 6 * (ar1_size + ar2_size)` calculation has no business of being guarded? Although you should (soon must maybe) make it `int(ar2_range)`.
Since you renamed the first variable, it doesn't even need to be in a `finally` clause.

Maybe someone else has an opinion.  I might just go with an error (you can still remove it). If  at all an option, it should be part of the kind like `"try-dictionary"`...

MilesCranmer replied:
Good point, will move it outside.

MilesCranmer replied:
Would something like
```python
        range_safe_from_overflow = ar2_range < np.iinfo(ar2.dtype).max
```
be what you had in mind? Then I can use:
```python
        ar2_range = int(ar2_max) - int(ar2_min)
```
which is cleaner as I don't have to use the `errstate`.

MilesCranmer replied:
Implemented this. Let me know what you think. I'll also add the requested unittests, etc.

MilesCranmer replied:
Unittests implemented. It should now test every error:
- `kind` is not of the three available.
- non-integral array passed but kind=dictionary.
- integer overflow but kind=dictionary.

It also tests lack of errors:
- integer overflow, but doesn't trigger error as kind=None or kind=mergesort.

seberg commented on numpy/lib/arraysetops.py:
```diff
@@ -593,6 +615,75 @@ def in1d(ar1, ar2, assume_unique=False, invert=False):
+                below_memory_constraint = False
+
+        # Use the fast integer algorithm
+        if below_memory_constraint or kind == 'dictionary':
```
This doesn't seem right.  If `kind == 'dictionary'` you ignore the overflow path above?  It seems there is a test missing for, `kind="dictionary"` when the overflow path would be hit?

MilesCranmer replied:
Good catch. Indeed, if the overflow happens with `ar2_max-ar2_min`, then the `ar2 - ar2_min` will overflow as well. Will fix.

seberg replied:
I was briefly considering if you should use Python integers rather than the `errstate`.  Then you still need to check against `np.iinfo(arr.dtype).max` though I guess.  So I am not sure it is better.

MilesCranmer replied:
Added.

seberg reviewed (commented):
Just two small comments, although I think this is very far along/good.

The one other ask I have is to add basic tests for all error paths. (incorrect "kind", use of the new method on float).

MilesCranmer committed:
MAINT: Protect against integer overflow in in1d

MilesCranmer committed:
MAINT: Clean up memory checking for in1d

MilesCranmer committed:
MAINT: Clean up integer overflow check in in1d

MilesCranmer committed:
MAINT: RuntimeError for unexpected integer overflow

MilesCranmer committed:
TST: validate in1d errors

MilesCranmer commented:
Thanks, all suggestions implemented! All error paths (and non-error paths, like integer overflow but `kind=None`) are now unittested. I also cleaned up the overflow check so it just compares `ar2_range` (now a Python int) with `np.iinfo(ar2.dtype).max` (@seberg's suggestion).

MilesCranmer committed:
DOC: Fix formatting issues in docstring

MilesCranmer committed:
DOC: Add missing indent in docstring

MilesCranmer committed:
DOC: Fix list format for sphinx

MilesCranmer commented:
Fixed some final sphinx formatting issues. This PR is all ready to go.

Let me know if anything else needs tweaks.

shoyer commented:
With regard to value names for the new `kind` parameter:

- How about sticking with the more generic "sort" rather than "mergesort"?
- "dictionary" seems a little misleading to me, since dictionaries refer to a specific type in Python, which we don't actually use here. How about using a more generic name like "table" instead?

MilesCranmer commented:
Great points, will implement. Thanks!

MilesCranmer committed:
MAINT: change kind names for in1d

- Switch dictionary->table, mergesort->sort

MilesCranmer commented:
Done.

MilesCranmer committed:
DOC: Improve clarity of in1d docstring

MilesCranmer committed:
DOC: `assume_unique` does not affect table method

MilesCranmer committed:
DOC: Notes on `kind` to in1d/isin docstring

MilesCranmer commented:
Anything else I need to do, or could this be merged?

Thanks,
Miles

seberg commented on numpy/lib/arraysetops.py:
```diff
@@ -544,6 +545,26 @@ def in1d(ar1, ar2, assume_unique=False, invert=False):
+        * If 'sort', will use a mergesort-based approach. This will have
+          a memory usage of roughly 6 times the sum of the sizes of
+          `ar1` and `ar2`, not accounting for size of dtypes.
+        * If 'table', will use a key-dictionary approach similar
```
If you can think of another word for `key-dictionary` that might be nice.  "lookup table"?

seberg replied:
It could be cool to shorten this and move some of the informatoin to the notes (even a table like the `np.sort` documentation).  But I think this is good for now.

seberg commented on numpy/lib/tests/test_arraysetops.py:
```diff
@@ -391,6 +542,40 @@ def test_in1d_with_arrays_containing_tuples(self):
+            #  if there is an integer overflow expected when computing the
+            #  range of ar2
+            assert_raises(
+                RuntimeError,
```
I am good with `RuntimeError` for now, just flagging in case anyone has an opinion.

seberg reviewed (commented):
Thanks, I think this is basically good to go (assuming nobody notes reservation on the API addition).

There are three main points:
1. This API should use `, *,`, there is no reason to allow passing `kind` positionally.
2. It may be better to fortify against booleans containin values other than `0` and `1`.  Yes it is weird (a test is to make a `uint8` arr and view it as bool.)
3. It would be nice to parametrize the tests and simplify them using that.  For the most part, this is good but a few tests duplicate things unnecessary I think.

But, overall this looks good now and most of my comments are nitpicky!  So I think this is basically at the finish line.

MilesCranmer committed:
MAINT: Minor suggestions from code review

Co-authored-by: Sebastian Berg <sebastian@sipsolutions.net>

MilesCranmer committed:
MAINT: Remove positionality of kind in isin

MilesCranmer committed:
DOC: Clean up change log

MilesCranmer committed:
DOC: Rephrase docstring of in1d/isin

MilesCranmer committed:
MAINT: Fix edgecase for bool containers

MilesCranmer commented on numpy/lib/tests/test_arraysetops.py:
```diff
@@ -391,6 +542,40 @@ def test_in1d_with_arrays_containing_tuples(self):
+        # Error 1: `kind` is not one of 'sort' 'table' or None.
+        ar1 = np.array([1, 2, 3, 4, 5])
+        ar2 = np.array([2, 4, 6, 8, 10])
+        assert_raises(ValueError, in1d, ar1, ar2, kind='quicksort')
```
`assert_raises` is used elsewhere in this test file (for tests not changed in this PR), so maybe we keep the consistency for now and change them all later?

MilesCranmer committed:
TST: Reduce code re-use with pytest mark

MilesCranmer commented on numpy/lib/tests/test_arraysetops.py:
```diff
@@ -326,6 +436,47 @@ def test_in1d_ravel(self):
+                           ec)
+        assert_array_equal(in1d(a, long_b, assume_unique=False,
+                                kind="sort"),
+                           ec)
```
Done.

MilesCranmer commented:
Thanks! All suggestions implemented.

seberg commented:
I am still curious if the `ar1 = ar1 + np.uint(8)` is actually necessary, but it doesn't matter.  Yes we can skip the `pytest.raises()` for now (I do think it is fine to mix it in a file).

The parametrization is causing some empty array tests to fail because `[]` is `np.array([])` which is float64 (and thus gets correctly rejected).  (Maybe more, but those are the ones I noticed.)

MilesCranmer committed:
TST: Skip empty arrays for kind="table"

MilesCranmer commented:
Thanks, fixed!

MilesCranmer commented:
Anything else left or could this be merged?

seberg commented:
Looks good to me now, lets give it a shot.  Thanks @MilesCranmer and everyone who reviewed!  Sorry it took a bit of iterating, new API makes things a bit slower to move.

It would be awesome to follow-up a bit on the tests especially, but not particularly important.

If anyone has concerns about the API addition or sees a fixup that would be good, please comment or open an issue!

MilesCranmer commented:
Awesome, thanks so much @seberg and everyone else who helped with the PR!

InessaPawson commented:
Hi-five on merging your first pull request to NumPy, @MilesCranmer! We hope you stick around!

WarrenWeckesser commented:
This might have introduced a bug; take a look at https://github.com/numpy/numpy/issues/21841

MilesCranmer commented:
Fixed on https://github.com/numpy/numpy/pull/21842. Mistake was not unit-testing for empty integer arrays–those are now added.

MilesCranmer commented:
Other post-merge question: is 'timedelta64' technically a subtype of `np.integer`? If so I think that might break things.

(I'm trying to make a similar method for `unique1d`, but it has a unittest for `timedelta64` which gets past the integer check)

adeak commented:
> How should I check for pure integer arrays (i.e., no datetimes or other dtypes which can't get evaluated by int())?

I'm not sure you can do much short of special-casing the "known bad" integer-looking dtypes. There's only this one built in NumPy.

This is what seems like a problem to me:
```py
>>> np.timedelta64.mro()
[numpy.timedelta64,
 numpy.signedinteger,
 numpy.integer,
 numpy.number,
 numpy.generic,
 object]

>>> np.int64.mro()
[numpy.int64,
 numpy.signedinteger,
 numpy.integer,
 numpy.number,
 numpy.generic,
 object]
```
`int64` and `timedelta64` are sibling classes, which means you don't have any superdtype for `int64` that doesn't include `timedelta64`. But it seems to be the only such edge case:
```py
>>> np.integer.__subclasses__()
[numpy.signedinteger, numpy.unsignedinteger]

>>> np.signedinteger.__subclasses__()
[numpy.int8,
 numpy.int16,
 numpy.int32,
 numpy.int64,
 numpy.longlong,
 numpy.timedelta64]

>>> np.unsignedinteger.__subclasses__()
[numpy.uint8, numpy.uint16, numpy.uint32, numpy.uint64, numpy.ulonglong]
```

It seems clear to me that dropping the pure-python `int` range check is not an option, we should just force the weird integer-like dtypes to use the old code path.

So one question remains: can users have custom integer-looking dtypes that don't work with `int`? If yes, we would have to manually list the "known good" integer types, plus `bool_`. If we're happy with breaking on that, we could just test for "`np.integer` but not `np.timedelta64`".

shoyer commented:
For what it's worth, timedelta64 definitely should not be a subclass of int. This can and should be changed.

I have an old PR to fix this that was never finished up if someone wants to help: https://github.com/numpy/numpy/pull/10701

The only real debate was whether timedelta64 and datetime64 should share a base class or not.

seberg commented:
> For what it's worth, timedelta64 definitely should not be a subclass of int. This can and should be changed.

Yes, but it is a change with somewhat unclear side-effects downstream?  In this case, I would prefer adding an explicit solution here for now (and test for datetime/timedelta).
